### PR TITLE
Spring_Tests_HW

### PR DESCRIPTION
**Pull request by AlexPravda1**
Implemented test cases for:
- security package
- validation package
- dao package
- service package

### CODE REVIEW

File: src/main/java/mate/academy/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -19,7 +19,7 @@ public UserDaoImpl(SessionFactory sessionFactory) {
     public Optional<User> findByEmail(String email) {
         try (Session session = sessionFactory.openSession()) {
             return session.createQuery(
-                    "from User u join fetch u.roles where u.email = :email", User.class)
+                    "from User u left join fetch u.roles where u.email = :email", User.class)
```
Comment (Nick97-git) [inline src/main/java/mate/academy/dao/impl/UserDaoImpl.java:22]: what for?

File: src/test/java/mate/academy/AbstractTest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                        .build()
+        );
+    }
+
+    protected Properties getProperties() {
+        Properties properties = new Properties();
+        properties.put("hibernate.dialect", getDataSourceProvider().hibernateDialect());
+        properties.put("hibernate.hbm2ddl.auto", "create-drop");
+
+        //data source settings
```
Comment (Nick97-git) [inline src/test/java/mate/academy/AbstractTest.java:74]: remove empty line and comment

File: src/test/java/mate/academy/AbstractTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import org.hibernate.Interceptor;
+import org.hibernate.SessionFactory;
+import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
+import org.hibernate.cfg.Configuration;
+import org.hsqldb.jdbc.JDBCDataSource;
+import org.junit.jupiter.api.BeforeEach;
+
+public abstract class AbstractTest {
+    protected interface DataSourceProvider {
+
```
Comment (Nick97-git) [inline src/test/java/mate/academy/AbstractTest.java:16]: remove empty line

File: src/test/java/mate/academy/dao/impl/RoleDaoImplTest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @BeforeEach
+    void setUp() {
+        roleDao = new RoleDaoImpl(getSessionFactory());
+        role = new Role(Role.RoleName.USER);
+    }
+
+    @Test
+    void save_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/RoleDaoImplTest.java:27]:
```suggestion
    void save_Ok() {
```

Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/RoleDaoImplTest.java:27]: let's add a test for the case if you try to add a role with a name that already exists in the database

File: src/test/java/mate/academy/dao/impl/RoleDaoImplTest.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        Role actual = roleDao.save(role);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(role.getRoleName(), actual.getRoleName());
+    }
+
+    @Test
+    void getRoleByName_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/RoleDaoImplTest.java:34]:
```suggestion
    void getRoleByName_Ok() {
```

Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/RoleDaoImplTest.java:34]: let's add a test for the case if a role with such name doesn't exist

File: src/test/java/mate/academy/dao/impl/UserDaoImplTest.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import mate.academy.model.Role;
+import mate.academy.model.User;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class UserDaoImplTest extends AbstractTest {
+    private static final String USER_EMAIL = "user@gmail.com";
+    private static final String USER_PASSWORD = "12345";
+    private static final Long USER_ID = 1L;
```
Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/UserDaoImplTest.java:15]: don't need to make constants from these values

File: src/test/java/mate/academy/dao/impl/UserDaoImplTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    @BeforeEach
+    void setUp() {
+        userDao = new UserDaoImpl(getSessionFactory());
+        user = new User();
+        user.setPassword(USER_PASSWORD);
+        user.setEmail(USER_EMAIL);
+    }
+
+    @Test
+    void findByEmail() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/UserDaoImplTest.java:33]:
```suggestion
    void findByEmail_Ok() {
```

Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/UserDaoImplTest.java:33]: add test for the case if user with such email doesn't exist

File: src/test/java/mate/academy/dao/impl/UserDaoImplTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        User savedUser = userDao.save(user);
+        Optional<User> actual = userDao.findByEmail(USER_EMAIL);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(savedUser.getEmail(), actual.get().getEmail());
+    }
+
+    @Test
+    void save_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/UserDaoImplTest.java:41]:
```suggestion
    void save_Ok() {
```

Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/UserDaoImplTest.java:41]: add a test for the case if a user with such email already added to the database

File: src/test/java/mate/academy/security/AuthenticationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import mate.academy.service.UserService;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+class AuthenticationServiceImplTest {
+    private static final String USER_EMAIL = "user@gmail.com";
+    private static final String USER_PASSWORD = "12345";
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/AuthenticationServiceImplTest.java:20]: don't need to make constants from these values

File: src/test/java/mate/academy/security/AuthenticationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                                                            roleService,
+                                                            passwordEncoder);
+        user = new User();
+        user.setEmail(USER_EMAIL);
+        user.setPassword(USER_PASSWORD);
+        user.setRoles(Set.of(new Role(Role.RoleName.USER)));
+    }
+
+    @Test
+    void register_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/AuthenticationServiceImplTest.java:42]:
```suggestion
    void register_Ok() {
```

File: src/test/java/mate/academy/security/AuthenticationServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        Mockito.when(roleService.getRoleByName("USER")).thenReturn(new Role(Role.RoleName.USER));
+
+        User actual = authenticationService.register(USER_EMAIL, USER_PASSWORD);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(USER_EMAIL, actual.getEmail());
+        Assertions.assertEquals(USER_PASSWORD, actual.getPassword());
+    }
+
+    @Test
+    void login_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/AuthenticationServiceImplTest.java:53]:
```suggestion
    void login_Ok() {
```

File: src/test/java/mate/academy/security/AuthenticationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            Assertions.assertEquals(USER_EMAIL, actual.getEmail());
+            Assertions.assertEquals(USER_PASSWORD, actual.getPassword());
+
+        } catch (AuthenticationException e) {
+            throw new RuntimeException("AuthenticationServiceImpl login test have failed.", e);
+        }
+    }
+
+    @Test
+    void login_NotOk() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/AuthenticationServiceImplTest.java:69]:
```suggestion
    void login_nonExistentUser_NotOk() {
```

File: src/test/java/mate/academy/security/AuthenticationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        Mockito.when(userService.findByEmail(USER_EMAIL))
+                .thenReturn(Optional.of(new User()));
+
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/AuthenticationServiceImplTest.java:72]: remove empty line

File: src/test/java/mate/academy/security/AuthenticationServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        try {
+            authenticationService.login(USER_EMAIL, USER_PASSWORD);
+        } catch (AuthenticationException e) {
+            Assertions.assertEquals("Incorrect username or password!!!", e.getMessage());
+            return;
+        }
+        Assertions.fail("Expected to receive AuthenticationException");
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/AuthenticationServiceImplTest.java:79]: https://howtodoinjava.com/junit5/expected-exception-example/

File: src/test/java/mate/academy/security/CustomUserDetailsServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+
+class CustomUserDetailsServiceTest {
+    private static final String USER_EMAIL = "user@gmail.com";
+    private static final String USER_PASSWORD = "12345";
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/CustomUserDetailsServiceTest.java:17]: the same as above

File: src/test/java/mate/academy/security/CustomUserDetailsServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        userService = Mockito.mock(UserService.class);
+        userDetailsService = new CustomUserDetailsService(userService);
+        user = new User();
+        user.setEmail(USER_EMAIL);
+        user.setPassword(USER_PASSWORD);
+        user.setRoles(Set.of(new Role(Role.RoleName.USER)));
+    }
+
+    @Test
+    void loadUserByUsername() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/CustomUserDetailsServiceTest.java:33]:
```suggestion
    void loadUserByUsername_Ok() {
```

Comment (Nick97-git) [inline src/test/java/mate/academy/security/CustomUserDetailsServiceTest.java:33]: add test for the case if email doesn't exist

File: src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+import org.mockito.Mockito;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.userdetails.User;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.test.util.ReflectionTestUtils;
+
+class JwtTokenProviderTest {
+    private static final String USER_EMAIL = "user@gmail.com";
+    private static final String USER_PASSWORD = "12345";
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java:19]: the same as above

File: src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    private static final String TOKEN_VALID = "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1c2VyQGdtYWlsLmNv"
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java:20]:
```suggestion
    private static final String VALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1c2VyQGdtYWlsLmNv"
```

File: src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    @Test
+    void createToken_OK() {
+        List<String> roles = List.of("USER");
+        String actual = jwtTokenProvider.createToken(USER_EMAIL, roles);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(TOKEN_VALID.length(), actual.length());
+    }
+
+    @Test
+    void getAuthentication() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java:43]:
```suggestion
    void getAuthentication_Ok() {
```

File: src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+        UserDetails userDetails = builder.build();
+        Mockito.when(this.userDetailsService.loadUserByUsername(any())).thenReturn(userDetails);
+        Authentication authentication = jwtTokenProvider.getAuthentication(TOKEN_VALID);
+        User actual = (User) authentication.getPrincipal();
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(actual.getUsername(), USER_EMAIL);
+    }
+
+    @Test
+    void getUsername() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java:57]:
```suggestion
    void getUsername_Ok() {
```

File: src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+        String actual = jwtTokenProvider.getUsername(TOKEN_VALID);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(actual, USER_EMAIL);
+    }
+
+    @Test
+    void resolveToken() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java:64]:
```suggestion
    void resolveToken_Ok() {
```

File: src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+        HttpServletRequest req = Mockito.mock(HttpServletRequest.class);
+        Mockito.when(req.getHeader("Authorization")).thenReturn("Bearer " + TOKEN_VALID);
+        String actual = jwtTokenProvider.resolveToken(req);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(actual, TOKEN_VALID);
+    }
+
+    @Test
+    void validateToken() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java:73]:
```suggestion
    void validateToken_validToken_Ok() {
```

Comment (Nick97-git) [inline src/test/java/mate/academy/security/jwt/JwtTokenProviderTest.java:73]: add test for the case if token is not valid

File: src/test/java/mate/academy/service/impl/RoleServiceImplTest.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+
+    @BeforeEach
+    void setUp() {
+        roleDao = Mockito.mock(RoleDao.class);
+        roleService = new RoleServiceImpl(roleDao);
+        role = new Role(Role.RoleName.USER);
+    }
+
+    @Test
+    void save_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/service/impl/RoleServiceImplTest.java:25]:
```suggestion
    void save_Ok() {
```

File: src/test/java/mate/academy/service/impl/RoleServiceImplTest.java [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+        Mockito.when(roleDao.save(role)).thenReturn(role);
+        Role actual = roleService.save(role);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(role.getRoleName(), actual.getRoleName());
+    }
+
+    @Test
+    void getRoleByName_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/service/impl/RoleServiceImplTest.java:33]:
```suggestion
    void getRoleByName_Ok() {
```

File: src/test/java/mate/academy/service/impl/UserServiceImplTest.java [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+class UserServiceImplTest {
+    private static final String USER_EMAIL = "user@gmail.com";
+    private static final String USER_PASSWORD = "12345";
+    private static final Long USER_ID = 1L;
```
Comment (Nick97-git) [inline src/test/java/mate/academy/service/impl/UserServiceImplTest.java:18]: the same as above

File: src/test/java/mate/academy/validation/EmailValidatorTest.java [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+        String email = "test@gmail.com";
+        Assertions.assertTrue(validator.isValid(email, constraintValidatorContext));
+        String emailShort = "i@i.com";
+        Assertions.assertTrue(validator.isValid(emailShort, constraintValidatorContext));
+        String emailLong = "testLongEmailLikeThis@gmailYahooSuperDomeain.com";
+        Assertions.assertTrue(validator.isValid(emailLong, constraintValidatorContext));
+    }
+
+    @Test
+    void isValid_NotOk() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/validation/EmailValidatorTest.java:30]:
```suggestion
    void isValid_invalidEmail_Ok() {
```

File: src/test/java/mate/academy/validation/EmailValidatorTest.java [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+    private ConstraintValidatorContext constraintValidatorContext;
+
+    @BeforeEach
+    void setUp() {
+        validator = new EmailValidator();
+        constraintValidatorContext = Mockito.mock(ConstraintValidatorContext.class);
+    }
+
+    @Test
+    void isValid_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/validation/EmailValidatorTest.java:20]:
```suggestion
    void isValid_validEmail_Ok() {
```

File: src/test/java/mate/academy/validation/PasswordValidatorTest.java [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+import javax.validation.ConstraintValidatorContext;
+import mate.academy.model.dto.UserRegistrationDto;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+class PasswordValidatorTest {
+    private static final String USER_EMAIL = "user@gmail.com";
+    private static final String USER_PASSWORD = "12345";
```
Comment (Nick97-git) [inline src/test/java/mate/academy/validation/PasswordValidatorTest.java:12]: the same as above

File: src/test/java/mate/academy/validation/PasswordValidatorTest.java [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+        registrationDto.setPassword(USER_PASSWORD);
+
+        constraintAnnotation = Mockito.mock(Password.class);
+        Mockito.when(constraintAnnotation.field()).thenReturn("password");
+        Mockito.when(constraintAnnotation.fieldMatch()).thenReturn("repeatPassword");
+        passwordValidator.initialize(constraintAnnotation);
+    }
+
+    @Test
+    void isValid_OK() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/validation/PasswordValidatorTest.java:33]:
```suggestion
    void isValid_validPassword_Ok() {
```

File: src/test/java/mate/academy/validation/PasswordValidatorTest.java [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+        registrationDto.setRepeatPassword(USER_PASSWORD);
+        Assertions.assertTrue(passwordValidator.isValid(
+                registrationDto, constraintValidatorContext), "Passwords should match");
+    }
+
+    @Test
+    void isValid_NotOk() {
```
Comment (Nick97-git) [inline src/test/java/mate/academy/validation/PasswordValidatorTest.java:40]:
```suggestion
    void isValid_invalidPassword_NotOk() {
```

**Review by Nick97-git — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/dao/impl/UserDaoImpl.java (hunk 1 above)
Comment (AlexPravda1) [reply src/main/java/mate/academy/dao/impl/UserDaoImpl.java:22]: REF to screens attached
1) Method fails if I try to save user with Set of Roles
2) findByEmail return Empty if I save without roles but leave just "join fetch" 
Any suggestions, please? 
Screenshot_2
Screenshot_1

Comment (Nick97-git) [reply src/main/java/mate/academy/dao/impl/UserDaoImpl.java:22]: You can't save user with role, that you doesn't have in DB, you should find this role in DB first and then set it in user

File: src/test/java/mate/academy/dao/impl/UserDaoImplTest.java [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+        User savedUser = userDao.save(user);
+        String userEmail = "user@gmail.com";
+        Optional<User> actual = userDao.findByEmail(userEmail);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(savedUser.getEmail(), actual.get().getEmail());
+    }
+
+    @Test
+    void findByEmail_nonExistentUser_Ok() {
+        userDao.save(user);
```
Comment (Nick97-git) [inline src/test/java/mate/academy/dao/impl/UserDaoImplTest.java:61]: do you need this line here?

**Review by Nick97-git — CHANGES_REQUESTED** [review]
Answered for your comment + you missed several comments from my previous review, you can see them in tab Files changed

File: src/test/java/mate/academy/dao/impl/UserDaoImplTest.java (hunk 31 above)
Comment (AlexPravda1) [reply src/test/java/mate/academy/dao/impl/UserDaoImplTest.java]: remade to make intentions more clear (to show that we are searching DB with clearly mismatching User details)

**Review by Nick97-git — APPROVED** [review]

**Pull request closed without merging**